torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -830,6 +830,113 @@ void removeMaxPoolUnusedOutput(Block* b) {
   }
 }
 
+// This optimization fuses LogSoftmax and NegativeLogLikelihoodLoss operators
+// into one operator: SoftmaxCrossEntropyLoss, and depending on the dimensions
+// of the input and different attributes there will be different subgraphs of
+// LogSoftmax and NegativeLogLikelihoodLoss.
+static void fuseLogSoftmaxNllLoss(Block* b) {
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    for (auto* child_block : it->blocks()) {
+      fuseLogSoftmaxNllLoss(child_block);
+    }
+    if (it->kind() == onnx::NegativeLogLikelihoodLoss) {
+      auto prev = it->input(0)->node();
+      Node* origNllLossNode = *it;
+      Node* origLogSoftmaxNode;
+      // if the input is 2D
+      // graph(%input : Float(3, 5),
+      // %target : Long(3)):
+      // %4 : Float(3, 5) = onnx::LogSoftmaxaxis=1
+      // %8 : Float(3) = onnx::NegativeLogLikelihoodLoss[reduction="none"]
+      // return (%8)
+      if (prev->kind() == onnx::LogSoftmax) {
+        origLogSoftmaxNode = it->input(0)->node();
+        // if the input is 4D
+        // graph(%input : Float(3, 5, 2, 7),
+        // %target : Long(3, 2, 7)):
+        // %4 : Tensor = onnx::Transpose[perm=[0, 3, 2, 1]] (%input)
+        // %5 : Tensor = onnx::LogSoftmax[axis=3] (%4)
+        // %6 : Float(3, 5, 2, 7) = onnx::Transpose[perm=[0, 3, 2, 1]] (%5)
+        // %10 : Float(3, 2, 7) =
+        // onnx::NegativeLogLikelihoodLoss[reduction="none"](%6, %target) return
+        // (%10)
+      } else if (
+          prev->kind() == onnx::Transpose &&
+          prev->input(0)->node()->kind() == onnx::LogSoftmax) {
+        origLogSoftmaxNode = prev->input(0)->node();
+        auto logSoftmaxNode = origLogSoftmaxNode->input(0)->node();
```

Comment:
nit: why is this node called logSoftmaxNode?

Reply:
it should be called transpose